### DISCUSSION TITLE
Proposal: Allow |= in object initializer syntax

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#8782] [Language Ideas]
@emsaks commented on [Sat Dec 22 2018](https://github.com/dotnet/roslyn/issues/32002)

It would be nice if you could use the `|=` operator within the object initializer syntax (i.e. `new myObj { Attributes |= myObjAttribute.SomeFlag }`).

This would enable adding a flag to existing flags set in the constructor.

[There's another proposal for extending the initializer syntax for event handlers [here](https://github.com/dotnet/roslyn/issues/5365).]

---

@YairHalberstadt commented on [Sun Dec 23 2018](https://github.com/dotnet/roslyn/issues/32002#issuecomment-449630499)

I would suggest you move this request to dotnet/csharplang where language requests are taken

**Comment by john-h-k** [comment]
While I see the idea behind this, I feel this is untrue to the point of an object initializer, which should be used for *initialization*, not extending previously set values. It seems more proper to simply OR it in another line

**Comment by emsaks** [comment]
@johnkellyoxford: re. simplicity, what if you're passing the object as an argument? As it stands you have to declare a holder variable, OR the flags on another line, and only then pass it (forces a one-liner lambda into a multiline and braces).

**Comment by john-h-k** [comment]
@emsaks yes it requires slightly more code, but initializer suggests initialization. Initializer isn't just a block of code that you can use to prevent having a multiline lambda. Maybe instead
* Refactor the constructor to allow passing of additional flags for ORing
* Use multiple lines, as initialization block shouldn't be used for mutating an object, it should be used for initializing it from nothing

**Comment by Thaina** [comment]
`&=` should also allowed too

**Comment by emsaks** [comment]
@johnkellyoxford : I feel you're being too pedantic about the word 'initializer'; *most* fields in a class already have a default value too (usually `0` or `default(T)`) and you change it to what you want. Flags are sort of an edge case that there may be a default flag set (say `UnprocessedFlag`) and you need to *add* another flag (i.e. `IsResidentFile`). The constructor is not always under your control and you don't always know what the default will be.

**Comment by john-h-k** [comment]
so add another line? This seems kind of pointless. It doesn't really solve a major problem, all it would do is change some multiline lambdas to single line lambdas

**Comment by iam3yal** [comment]
dup #109.

**Comment by jnm2** [comment]
@johnkellyoxford 

> While I see the idea behind this, I feel this is untrue to the point of an object initializer, which should be used for initialization, not extending previously set values.

What's the conceptual difference between adding a flag to the default flags and adding an item to the default items which is possible in today's syntax?

```c#
var x = new Foo
{
   Items = { addedItem },
   ExistingObject = { ChangedProperty = 42 }
};
```

**Comment by john-h-k** [comment]
@jnm2 it does fit that for consistency, however I dislike that conceptually too honestly. I only use object initializer to initialize things that aren't initialized by the constructor, and then mutate things as I want post initializer

**Comment by iam3yal** [comment]
> so add another line? This seems kind of pointless. It doesn't really solve a major problem, all it would do is change some multiline lambdas to single line lambdas

Not all features need to solve a major problem <s>or even a problem at all</s>.

**Comment by theunrepentantgeek** [comment]
> Not all features need to solve a major problem or even a problem at all.

😮

Every feature brings with it a cognitive load on every developer using C#.

Every feature needs to be designed, implemented, tested and maintained - by a team that has limited resources (because _every_ team has limited resources).

This is why [-100 points](https://blogs.msdn.microsoft.com/ericgu/2004/01/12/minus-100-points/) has practically reached meme status.

If a feature isn't solving an actual problem, why bother?

_Note: I'm not expressing an opinion on the original post in this thread._

**Comment by iam3yal** [comment]
@theunrepentantgeek I don't disagree, just poor wording on my part but what I meant is some features are added to the language and they are more a niche or a convenience than necessity and imo this feature falls into the niche category.
